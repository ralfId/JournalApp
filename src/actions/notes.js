import Swal from "sweetalert2";


import { db } from "../firebase/firebaseConfig";
import { loadNotes } from "../helpers/loadNotes";
import { uploadImage } from "../helpers/uploadImage";
import { types } from "../types/types";

export const startNewNote = () => {

    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const docRef = await db.collection(`${uid}/journal/notes`).add(newNote)
                .catch(e => {
                    Swal.fire('Saved', e.message, 'success');
                });
            dispatch(activeNote(docRef.id, newNote));
            dispatch(addNewNote(docRef.id, newNote));
        } catch (error) {
            console.log('startNewNote>>>>>>>>>>>>>>>>>' + error)
        }
    }
}

export const activeNote = (id, note) => ({
    type: types.notesSetActiveNote,
    payload: { id, ...note }
})

export const addNewNote = (id, note) => ({
    type: types.notesAddNewNote,
    payload: { id, ...note }
})


export const startLoadingNotes = (id) => {
    return async (dispatch) => {
        try {
            const notes = await loadNotes(id);
            dispatch(setNotes(notes));
        } catch (error) {
            console.log('startLoadingNotes >>>>>>' + error)
        }
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoadNotes,
    payload: notes
})

export const startSavingNote = (note) => {

    return async (dispatch, getState) => {

        const { uid } = getState().auth;

        if (!note.imageUrl) {
            delete note.imageUrl
        }

        const noteToFireStore = { ...note };
        delete noteToFireStore.id;

        await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFireStore)
            .catch(e => {
                console.log(e.message)
                Swal.fire('Saved', e.message, 'success');
            });
        dispatch(refreshUpdatedNote(note.id, noteToFireStore));


        Swal.fire({
            position: 'top-end',
            icon: 'success',
            title: 'Your note has been saved',
            text: `${note.title}`,
            showConfirmButton: false,
            timer: 1500
        })
    }
}

export const refreshUpdatedNote = (id, note) => ({
    type: types.notesRefrshUpdatedNote,
    payload: { id: id, note: { id, ...note } }
})

export const startUploadingImage = (file) => {
    return async (dispatch, getState) => {

        try {
            const activeNote = getState().notes.active;

            Swal.fire({
                title: 'Loading...',
                text: 'Please wait...',
                allowOutsideClick: false,
                showConfirmButton: false,
                didOpen: () => { Swal.showLoading() }
            })
            
            const imageFileUrl = await uploadImage(file);
            activeNote.imageUrl = imageFileUrl;

            dispatch(startSavingNote(activeNote));

            Swal.close();
        } catch (error) {
            console.log('startUploadingImage >>>>>>>>>>>>' + error)
        }

    }
}

export const startDeleteNote = (id) => {
    return async (dispath, getState) => {

        const uid = getState().auth.uid;

        Swal.fire({
            title: 'Loading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            didOpen: () => { Swal.showLoading() }
        })
        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispath(deleteNote(id));

        Swal.close()

    }
}

export const deleteNote = (id) => ({
    type: types.notesDeleteNote,
    payload: { id }
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaningNotes
})
