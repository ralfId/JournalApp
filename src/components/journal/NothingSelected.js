import React from 'react'

export const NothingSelected = () => {
    return (
        <div className="nothing__main-content animate__animated animate__fadeIn  animate__faster  ">
            <p>Select somethin, or create a new entry</p>
            <i className="far fa-star fa-4x mt-5"></i>
        </div>
    )
}
