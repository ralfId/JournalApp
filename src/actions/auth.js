import { firebase, googleAuthProvider } from '../firebase/firebaseConfig'
import Swal from 'sweetalert2'

import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import { notesLogout } from './notes';


export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch(startLoading());

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {

                dispatch(login(user.uid, user.displayName));

                dispatch(finishLoading());

            })
            .catch(e => {
                
                dispatch(finishLoading());

                Swal.fire({
                    icon: 'error',
                    title: '',
                    text: e.message,
                  })
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {

        firebase.auth().createUserWithEmailAndPassword(email, password)
        
            .then(async ({ user }) => {

                await user.updateProfile({ displayName: name })
                dispatch(login(user.uid, user.displayName))

            }).catch(error =>{
                 Swal.fire('Error', error.message, 'error')
                })
    }
}


export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => { dispatch(login(user.uid, user.displayName)) })
    }
}


export const login = (uid, displayName) => (
    {
        type: types.login,
        payload: { uid, displayName }
    })

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout())
        dispatch(notesLogout());
    }
}
export const logout = (params) => ({ type: types.logout})

