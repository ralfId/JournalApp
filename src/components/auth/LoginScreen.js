import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import validator from 'validator'


import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { removeError, setError } from '../../actions/ui'
import { useForm } from '../../hooks/useForms'

export const LoginScreen = () => {

    const dispatch = useDispatch();
    const { msgError, loading } = useSelector(state => state.ui)
    

    const [formValues, handleInputChange] = useForm({ email: 'ralf_raid@yopmail.com', password: '123456' })
    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startLoginEmailPassword(email, password))
        }
    }

    const isFormValid = () => {
        if (!validator.isEmail(email)) {
            dispatch(setError('email is not valid'))
            return false
        } else if (password.length < 5) {

            dispatch(setError('password should be at least 6 characters'))
            return false;
        }

        dispatch(removeError())
        return true;

    }


    const handleGoogleLogin = () => { dispatch(startGoogleLogin()) }


    return (
        <>
            <h3 className="auth__title"> Login Screen </h3>
            <form onSubmit={handleLogin} className="animate__animated animate__fadeIn  animate__faster  ">
                {
                    msgError &&
                    (
                        <div className="auth__alert-error">
                            {msgError}
                        </div>
                    )
                }
                <label htmlFor='email'>Email</label>
                <input className="auth__input" name='email' type='text' autoComplete='off' value={email} onChange={handleInputChange} />

                <label htmlFor='password'>Password</label>
                <input className="auth__input" name='password' type='password' autoComplete='off' value={password} onChange={handleInputChange} />
                <br />
                <button className="btn btn-primary btn-block" disabled={loading}>
                    SigIn
                </button>



                <div className="auth__social-networks">
                    <p>Login with social netwoks</p>
                    <div className="google-btn" onClick={handleGoogleLogin}>
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link className="link" to="/auth/register">
                    Crate new acount
                </Link>
            </form>
        </>
    )
}
