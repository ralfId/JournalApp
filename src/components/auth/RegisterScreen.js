import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import validator from 'validator'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'
import { setError, removeError } from '../../actions/ui'
import { useForm } from '../../hooks/useForms'

export const RegisterScreen = () => {


    const dispatch = useDispatch();
    const {msgError} = useSelector(st => st.ui);

    const [formValues, handleInputChange] = useForm(
        {
            name: 'rafael',
            email: 'ralf_raid@yopmail.com',
            password: '123456',
            confirmpassword: '123456',
        }
    )

    const { name, email, password, confirmpassword } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }
    }

    const isFormValid = () => {

        if (name.trim().length === 0) {

            dispatch(setError('name is required'))
            return false;

        } else if (!validator.isEmail(email)) {

            dispatch(setError('email is not valid'))
            return false;

        } else if (password !== confirmpassword || password.length < 5) {

            dispatch(setError('password should be at least 6 characters an match each other'))
            return false;

        }

        dispatch(removeError())
        return true;
    }


    return (
        <>
            <h3 className="auth__title"> Register Screen </h3>
            <form onSubmit={handleRegister} className="animate__animated animate__fadeIn  animate__faster  ">
                {
                    msgError &&
                    (
                        <div className="auth__alert-error">
                            {msgError}
                        </div>
                    )
                }
                <label htmlFor='name'>Name</label>
                <input className="auth__input" name='name' value={name} type='text' autoComplete='off' onChange={handleInputChange} />

                <label htmlFor='email'>Email</label>
                <input className="auth__input" name='email' value={email} type='text' autoComplete='off' onChange={handleInputChange} />

                <label htmlFor='password'>Password</label>
                <input className="auth__input" name='password' value={password} type='password' autoComplete='off' onChange={handleInputChange} />

                <label htmlFor='password'>Confirm password</label>
                <input className="auth__input" name='confirmpassword' value={confirmpassword} type='password' autoComplete='off' onChange={handleInputChange} />
                <br />
                <button className="btn btn-primary btn-block mb-5 mt-5">
                    Register
            </button>




                <Link className="link " to="/auth/login">
                    Already register?
            </Link>
            </form>
        </>
    )
}
