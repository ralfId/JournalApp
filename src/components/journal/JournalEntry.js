import React from 'react';
import moment from 'moment'
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({ id, title, body, date, imageUrl }) => {

    const dispatch = useDispatch();
    const noteDate = moment(date);
    
    const handleSelectNote = () => {
        dispatch(activeNote(id, {title, body, date, imageUrl}))
    }
    

    return (
        <div className="journal__entry pointer animate__animated animate__fadeIn  animate__faster  " onClick={ handleSelectNote }>
            {
                (imageUrl) &&
                <div className="journal__entry-picture"
                    style={{
                        backgroundImage: `url(${imageUrl})`,
                        backgroundSize: 'cover'
                    }}>

                </div>
            }
            <div className="journal__entry-body">
                <div className="journal__entry-title">
                    {title}
                </div>
                <div className="journal__entry-content">
                    {body}
                </div>
            </div>
            <div className="journal__date">
                <span>{noteDate.format('dddd')}</span>
                <h4>{ noteDate.format('Do')}</h4>
            </div>
        </div>
    )
}
