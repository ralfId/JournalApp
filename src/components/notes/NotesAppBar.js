import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSavingNote, startUploadingImage } from '../../actions/notes';

export const NotesAppBar = () => {


    const dispatch = useDispatch();

    const { active } = useSelector(state => state.notes)

    const handleSaveNote = () => {
        // console.log(active)
        dispatch(startSavingNote(active));
    }

    const handleSelectImage = () => {
        document.querySelector('#fileSelector').click()
    }
    
    const handleOpenFileSelector = (e) => {
        const file  = e.target.files[0]
        dispatch(startUploadingImage(file));
    }
    

    return (
        <div className="notes__appbar ">
            <span> 28 de agosto de 2021</span>
            <input id='fileSelector' type="file" onChange={handleOpenFileSelector} style={{display: 'none'}}/>
            <div>
                <button className="btn" onClick={handleSelectImage}>
                    Picture
                </button>
                <button className="btn" onClick={handleSaveNote}>
                    Save
                </button>
            </div>
        </div>
    )
}
