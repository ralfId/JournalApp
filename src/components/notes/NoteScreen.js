import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleteNote } from '../../actions/notes'
import { useForm } from '../../hooks/useForms'
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const { active: note } = useSelector(state => state.notes)
    const [formValues, handleInputChange, reset] = useForm(note);
    const { id, title, body } = formValues;

    const activeId = useRef(note.id);

    useEffect(() => {

        if (note.id !== activeId.current) {
            reset(note);
            activeId.current = note.id
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch(activeNote(formValues.id, { ...formValues }))
    }, [formValues, dispatch]);

    const handleDeleteNote = () => {
        dispatch(startDeleteNote(id));
    }
    
    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">

                <label className="mt-5 mb-1" htmlFor="title">Title</label>
                <input type="text"
                    placeholder="Title" name="title"
                    className="notes__title-input" value={title} onChange={handleInputChange} />

                <label className="mt-5 mb-1" htmlFor="body">Description</label>
                <textarea className="notes__textarea " name="body"
                    placeholder="What happened today?"
                    value={body} onChange={handleInputChange}></textarea>

                <div className="notes__image">
                    {
                        (note.imageUrl)
                        &&
                        (
                            <img src={note.imageUrl} alt="fantastic" />
                        )
                    }
                </div>
            </div>

            <button className="btn btn-danger" onClick={handleDeleteNote}>
                 Delete
            </button>

        </div>

    )
}
