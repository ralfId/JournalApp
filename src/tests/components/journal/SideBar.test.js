import React from 'react';
import { mount } from "enzyme";
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'

import '@testing-library/jest-dom';

import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';
import { SideBar } from '../../../components/journal/SideBar';


jest.mock('../../../actions/auth', () => (({
    startLogout: jest.fn()
})))
jest.mock('../../../actions/notes', () => ({
    startNewNote: jest.fn()
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)


const initState = {
    auth: {
        uid: '1',
        displayName: 'Rafael'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: null
    }

};

let store = mockStore(initState)
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <SideBar />
    </Provider>
)


describe('pruebas en SideBar.js', () => {

    test('debe de mostrarse correctamente <SideBar/> ', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('debe de llamar el startLogout()', () => {
       
        wrapper.find('button').prop('onClick')();
        expect(wrapper.find('button').exists()).toBe(true)

    })

    test('debe de llamar el startNewNote()', () => {

        wrapper.find('.journal__new-entry').prop('onClick')();
        expect(startNewNote).toHaveBeenCalled()
    })



});