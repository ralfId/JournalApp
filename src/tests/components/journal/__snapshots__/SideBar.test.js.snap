// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en SideBar.js debe de mostrarse correctamente <SideBar/>  1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__side-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
          
         
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-3x"
    />
    <p
      className="mt-1"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries animate__animated animate__fadeIn  animate__faster  "
  />
</aside>
`;
